clarity/src/vm/tests/variables.rs

```diff
@@ -221,3 +221,466 @@ fn test_tenure_height(
+    );
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_stacks_block_height(
+    version: ClarityVersion,
+    epoch: StacksEpochId,
+    mut tl_env_factory: TopLevelMemoryEnvironmentGenerator,
+) {
+    // data var
+    expect_contract_error(
+        version,
+        epoch,
+        &mut tl_env_factory,
+        "data-var",
+        r#"
+        (define-data-var stacks-block-height uint u1234)
+        (define-read-only (test-func)
+            (var-get stacks-block-height)
+        )
+        "#,
+        &[(
+            WhenError::Initialization,
+            |version, _| version >= ClarityVersion::Clarity3,
+            CheckErrors::NameAlreadyUsed("stacks-block-height".to_string()),
+        )],
+        Value::UInt(1234),
+    );
+
+    // map
+    expect_contract_error(
+        version,
+        epoch,
+        &mut tl_env_factory,
+        "map",
+        r#"
+        (define-map stacks-block-height uint uint)
+        (define-private (test-func)
+            (map-insert stacks-block-height u1 u2)
+        )
+        "#,
+        &[(
+            WhenError::Initialization,
+            |version, _| version >= ClarityVersion::Clarity3,
+            CheckErrors::NameAlreadyUsed("stacks-block-height".to_string()),
+        )],
+        Value::Bool(true),
+    );
+
+    // let
+    expect_contract_error(
+        version,
+        epoch,
+        &mut tl_env_factory,
+        "let",
+        r#"
+        (define-private (test-func)
+            (let ((stacks-block-height 32))
+                stacks-block-height
+            )
+        )
+        "#,
+        &[(
+            WhenError::Runtime,
+            |version, _| version >= ClarityVersion::Clarity3,
+            CheckErrors::NameAlreadyUsed("stacks-block-height".to_string()),
+        )],
+        Value::Int(32),
+    );
+
+    // match binding
+    expect_contract_error(
+        version,
+        epoch,
+        &mut tl_env_factory,
+        "match-binding",
+        r#"
+        (define-read-only (test-func)
+          (let ((x (if true (ok u5) (err u7))))
+            (match x
+              stacks-block-height 3
+              e 4
+            )
+          )
+        )
+        "#,
+        &[(
+            WhenError::Runtime,
+            |version, _| version >= ClarityVersion::Clarity3,
+            CheckErrors::NameAlreadyUsed("stacks-block-height".to_string()),
+        )],
+        Value::Int(3),
+    );
+
+    // function
+    expect_contract_error(
+        version,
+        epoch,
+        &mut tl_env_factory,
+        "function",
+        r#"
+        (define-private (stacks-block-height) true)
+        (define-private (test-func) (stacks-block-height))
+        "#,
+        &[(
+            WhenError::Initialization,
+            |version, _| version >= ClarityVersion::Clarity3,
+            CheckErrors::NameAlreadyUsed("stacks-block-height".to_string()),
+        )],
+        Value::Bool(true),
+    );
+}
+
+#[apply(test_clarity_versions)]
+fn reuse_tenure_height(
```

Comment:
Also tagging this for testing with all Clarity built-ins for creating named data.

Reply:
Added in 890f07cd1d2b939f9f9daa0f497824e7b4359ff7.